blockcache/blockcache_test.go

```diff
@@ -0,0 +1,190 @@
+	require.Equal(t, 1, mc.chainCallCount)
+	mc.resetChainCallCount()
+
+	_, err = bc.Cache.Get(*inv1)
+	require.NoError(t, err)
+
+	// After calling getBlock for block2, it is expected that the Cache
+	// will have a size of 2 and will contain both block1 and block2.
+	// One chain backends call is expected to fetch the block.
+	_, err = bc.GetBlock(&blockhash2, getBlockImpl)
+	require.NoError(t, err)
+	require.Equal(t, 2, bc.Cache.Len())
+	require.Equal(t, 1, mc.chainCallCount)
+	mc.resetChainCallCount()
+
+	_, err = bc.Cache.Get(*inv1)
+	require.NoError(t, err)
+
+	_, err = bc.Cache.Get(*inv2)
+	require.NoError(t, err)
+
+	// getBlock is called again for block1 to make block2 the LFU block.
+	// No call to the chain backend is expected since block 1 is already
+	// in the Cache.
+	_, err = bc.GetBlock(&blockhash1, getBlockImpl)
+	require.NoError(t, err)
+	require.Equal(t, 2, bc.Cache.Len())
+	require.Equal(t, 0, mc.chainCallCount)
+	mc.resetChainCallCount()
+
+	// Since the Cache is now at its max capacity, it is expected that when
+	// getBlock is called for a new block then the LFU block will be
+	// evicted. It is expected that block2 will be evicted. After calling
+	// Getblock for block3, it is expected that the Cache will have a
+	// length of 2 and will contain block 1 and 3.
+	_, err = bc.GetBlock(&blockhash3, getBlockImpl)
+	require.NoError(t, err)
+	require.Equal(t, 2, bc.Cache.Len())
+	require.Equal(t, 1, mc.chainCallCount)
+	mc.resetChainCallCount()
+
+	_, err = bc.Cache.Get(*inv1)
+	require.NoError(t, err)
+
+	_, err = bc.Cache.Get(*inv2)
+	require.True(t, errors.Is(err, cache.ErrElementNotFound))
+
+	_, err = bc.Cache.Get(*inv3)
+	require.NoError(t, err)
+
+	require.Len(t, bc.blockMtxs.mutexMap, 0)
+}
+
+// TestBlockCacheMutexs is used to test that concurrent calls to GetBlock with
+// the same block hash does not result in multiple calls to the chain backend.
+// In other words this tests the blockMtxs.
+func TestBlockCacheMutexs(t *testing.T) {
+	mc := newMockChain()
+	getBlockImpl := mc.GetBlock
+
+	block1 := &wire.MsgBlock{Header: wire.BlockHeader{Nonce: 1}}
+	block2 := &wire.MsgBlock{Header: wire.BlockHeader{Nonce: 2}}
+
+	blockhash1 := block1.BlockHash()
+	blockhash2 := block2.BlockHash()
+
+	// Determine the size of the block.
+	sz, _ := (&cache.CacheableBlock{Block: btcutil.NewBlock(block1)}).Size()
+
+	// A new Cache is set up with a capacity of 2 blocks
+	bc := NewBlockCache(2 * sz)
+
+	mc.addBlock(&wire.MsgBlock{}, 1)
+	mc.addBlock(&wire.MsgBlock{}, 2)
+
+	// Spin off multiple go routines and ensure that concurrent calls to the
+	// GetBlock method does not result in multiple calls to the chain
+	// backend.
+	var wg sync.WaitGroup
+	for i := 0; i < 100; i++ {
+		wg.Add(1)
+		go func(e int) {
+			if e%2 == 0 {
+				_, err := bc.GetBlock(&blockhash1, getBlockImpl)
+				require.NoError(t, err)
+			} else {
+				_, err := bc.GetBlock(&blockhash2, getBlockImpl)
+				require.NoError(t, err)
```

Comment:
`require` package has some [strange behaviour](https://github.com/stretchr/testify/issues/524) when it's called in goroutines, because it [calls](https://github.com/stretchr/testify/blob/6990a05d54c2287be8787a411d77815643347339/require/require.go#L1264) `t.FailNow` it should [only be used the main test goroutine](https://golang.org/pkg/testing/#T). I haven't actually run into this in practice, but it's probably better to just use `assert.NoError` in this goroutine in case - it only calls [fail](https://github.com/stretchr/testify/blob/6990a05d54c2287be8787a411d77815643347339/assert/assertions.go#L1332).